File: python/lld/cache-design/key-value-store/kv_store.py
import time
from entry import Entry
from expiry_queue import ExpiryQueue


class StaleWrite(Exception):
    pass


class KVStore:
    """The map + operations. The lazy check on get() backstops the
    proactive Sweeper — a key is never returned past its expiry,
    whichever runs first."""

    def __init__(self):
        self._entries = {}
        self.expiry_queue = ExpiryQueue()

    def set(self, key: str, value, ttl_ms: float) -> None:
        expires_at = time.time() * 1000 + ttl_ms
        prev_version = self._entries[key].version if key in self._entries else 0
        self._entries[key] = Entry(value, expires_at, prev_version + 1)
        self.expiry_queue.schedule(key, expires_at)

    def get(self, key: str):
        entry = self._entries.get(key)
        if entry is None:
            return None
        if time.time() * 1000 > entry.expires_at:   # lazy check
            del self._entries[key]
            return None
        return entry.value

    def delete(self, key: str) -> None:
        self._entries.pop(key, None)

    def compare_and_set(self, key: str, expected_version: int, value, ttl_ms: float) -> None:
        current = self._entries.get(key)
        current_version = current.version if current else 0
        if current_version != expected_version:
            raise StaleWrite()
        self.set(key, value, ttl_ms)

    def peek_live(self, key: str):
        return self._entries.get(key)

    def force_delete(self, key: str) -> None:
        self._entries.pop(key, None)

File: java/lld/cache-design/key-value-store/KVStore.java
import java.util.HashMap;
import java.util.Map;

public class KVStore<V> {

    public static class StaleWrite extends RuntimeException {}

    private final Map<String, Entry<V>> entries = new HashMap<>();
    public final ExpiryQueue expiryQueue = new ExpiryQueue();

    // The map + operations. The lazy check on get() backstops the
    // proactive Sweeper — a key is never returned past its expiry,
    // whichever runs first.
    public void set(String key, V value, long ttlMs) {
        long expiresAt = System.currentTimeMillis() + ttlMs;
        int prevVersion = entries.containsKey(key) ? entries.get(key).version : 0;
        entries.put(key, new Entry<>(value, expiresAt, prevVersion + 1));
        expiryQueue.schedule(key, expiresAt);
    }

    public V get(String key) {
        Entry<V> entry = entries.get(key);
        if (entry == null) return null;
        if (System.currentTimeMillis() > entry.expiresAt) { entries.remove(key); return null; }   // lazy check
        return entry.value;
    }

    public void delete(String key) { entries.remove(key); }

    public void compareAndSet(String key, int expectedVersion, V value, long ttlMs) {
        Entry<V> current = entries.get(key);
        int currentVersion = current == null ? 0 : current.version;
        if (currentVersion != expectedVersion) throw new StaleWrite();
        set(key, value, ttlMs);
    }

    public Entry<V> peekLive(String key) { return entries.get(key); }
    public void forceDelete(String key) { entries.remove(key); }
}